pilot/pkg/proxy/envoy/v2/mesh.go

```diff
@@ -0,0 +1,990 @@
+}
+
+// SubsetNames implements functionality required for CDS and returns a list of all subset
+// names currently configured for this Mesh
+func (m *Mesh) SubsetNames() []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	out := make([]string, len(m.subsetEndpoints))
+	idx := 0
+	for ssName := range m.subsetEndpoints {
+		out[idx] = ssName
+		idx++
+	}
+	return out
+}
+
+// UpdateRules implements functionality required for pilot configuration (via Galley).
+// It updates the Mesh for supplied events, adding, updating and deleting destination
+// rules from this mesh as determined by the corresponding Event.
+func (m *Mesh) UpdateRules(ruleChanges []RuleChange) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	var errs error
+	for _, ruleChange := range ruleChanges {
+		rule := ruleChange.Rule
+		ruleType, labelValue, cidrNet := determineRuleType(rule.Name)
+		for _, subset := range rule.Subsets {
+			scopedSSName := rule.Name + RuleSubsetSeparator + subset.Name
+			if ruleChange.Type == ConfigDelete || ruleChange.Type == ConfigUpdate {
+				oldEpSet := m.subsetEndpoints[scopedSSName]
+				m.reverseEpSubsets.deleteSubsetForEndpoints(oldEpSet, scopedSSName)
+				delete(m.subsetEndpoints, scopedSSName)
+				delete(m.subsetDefinitions, scopedSSName)
+			}
+			if ruleChange.Type == ConfigAdd || ruleChange.Type == ConfigUpdate {
+				var labelsToMatch map[string]string
+				switch ruleType {
+				case DestinationRuleService:
+					labelsToMatch =
+						newMapWithLabelValue(subset.GetLabels(), DestinationService.AttrName(), labelValue)
+				case DestinationRuleIP:
+					labelsToMatch =
+						newMapWithLabelValue(subset.GetLabels(), DestinationIP.AttrName(), labelValue)
+				case DestinationRuleName:
+					labelsToMatch =
+						newMapWithLabelValue(subset.GetLabels(), DestinationName.AttrName(), labelValue)
+				case DestinationRuleCIDR | DestinationRuleWildcard:
+					// Direct matches will not work. Scoping is done later.
+					labelsToMatch = subset.GetLabels()
+				}
+				newEpSubset := m.reverseAttrMap.getEndpointsMatching(labelsToMatch)
+				if ruleType == DestinationRuleCIDR || ruleType == DestinationRuleWildcard {
+					// For rule names that do not map to values that can be label matched
+					// scope the returned set to the type of rule.
+					newEpSubset = newEpSubset.scopeToRule(ruleType, labelValue, cidrNet)
+				}
+				m.subsetEndpoints.addEndpointSet(scopedSSName, newEpSubset)
+				m.reverseEpSubsets.addSubsetForEndpoints(newEpSubset, scopedSSName)
+				m.subsetDefinitions[scopedSSName] = subset
+			}
+		}
+	}
+	return errs
+}
+
+// NewEndpoint is a boiler plate function intended for platform Controllers to create a new Endpoint.
+// This method ensures all the necessary data required for creating subsets are correctly setup. It
+// also performs sorting of arrays etc, to allow stable results for reflect.DeepEquals() for quick
+// comparisons. The address	is the network address of this endpoint that must be accessible from __THIS__
```

Comment:
nit: extra spaces in this comment block due to old line breaks.

Reply:
done